### PR TITLE
types: add typescript definitions for unified-engine

### PR DESCRIPTION
**Pull request by ChristianMurphy**
TODO:
- [x] add documentation comments to options
- [x] add complete type tests
- [x] Look into XO warnings

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    ignorePatterns?: string[]
+
+    silentlyIgnore?: boolean
+
+    plugins?: Pluggable<[Settings?], P> | PluggableList<P>
+
+    pluginsPrefix?: string
+
+    configTransform?: (config: unknown) => Options<P>
```
Comment (ChristianMurphy) [inline types/index.d.ts]: @wooorm could use some clarity on this.
The example shows only `settings` being overriden, but it looks like it can override more.
Currently I have it set to be able to override any attribute in unified engine options, is this accurate?

File: types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+     */
+    processor: Processor<P>
+
+    cwd?: string
+
+    files?: Array<string | VFile>
+
+    extensions?: string[]
+
+    streamIn?: ReadStream
```
Comment (Rokt33r) [inline types/index.d.ts]: Should be `ReadableStream`

File: types/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    filePath?: string
+
+    streamOut?: WriteStream
```
Comment (Rokt33r) [inline types/index.d.ts]: `WritableStream`

File: types/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    streamError?: WriteStream
```
Comment (Rokt33r) [inline types/index.d.ts]: WritableStream

File: types/index.d.ts (hunk 2 above)
Comment (ChristianMurphy) [reply types/index.d.ts]: Good catch, updated

File: types/index.d.ts (hunk 3 above)
Comment (ChristianMurphy) [reply types/index.d.ts]: Good catch, updated

File: types/index.d.ts (hunk 4 above)
Comment (ChristianMurphy) [reply types/index.d.ts]: Good catch, updated

File: types/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,275 @@
+// TypeScript Version: 3.0
+
+import {Plugin, Pluggable, PluggableList, Processor, Settings} from 'unified'
+import {VFile} from 'vfile'
+
+declare namespace unifiedEngine {
+  type VFileReporter<T = Settings> = (files: VFile[], options: T) => string
```
Comment (ChristianMurphy) [inline types/index.d.ts]: @Rokt33r Not sure if it would make sense to upstream this to vfile or another repo or keep it here.
Currently leaning towards keeping it here, thoughts? :thought_balloon:

File: types/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+  }
+
+  interface FileSet {
+    files: VFile[]
+    origins: string[]
+    expected: number
+    actual: number
+    pipeline: unknown
+    plugins: Plugin[]
+  }
```
Comment (ChristianMurphy) [inline types/index.d.ts]: This is extracted from reading the source,
not entirely sure on the type for `pipeline`, would it be a `Processor` in this context? or a lower level trough? (this could be kept as unknown for now and added in a later minor/patch release)
Also open on descriptions I have a decent guess of what `files`, `origins`, and `plugins` are intended to do.
less clear on `expected`, `actual` and `pipeline`'s intended use.

File: package.json [hunk 7]
```diff
@@ -11,12 +11,19 @@
   "repository": "unifiedjs/unified-engine",
   "bugs": "https://github.com/unifiedjs/unified-engine/issues",
   "author": "Titus Wormer <tituswormer@gmail.com> (https://wooorm.com)",
+  "funding": {
+    "type": "opencollective",
+    "url": "https://opencollective.com/unified"
+  },
```
Comment (ChristianMurphy) [inline package.json]: Also included funding field, this could also be split into another PR, if preferred.

File: types/index.d.ts (hunk 1 above)
Comment (wooorm) [reply types/index.d.ts]: `configTransform` is given the result from a config file, such as `packagejson#packageField`, `config.json`, `config.yml`, or `config.js`, and must return a value that is expected by the unified engine (a preset). One example of this is alex, which has custom config files: https://github.com/get-alex/alex/blob/2292d1a8a12e71b2deb7ced3a635f61842bd1927/cli.js#L120-L146

File: types/index.d.ts (hunk 6 above)
Comment (wooorm) [reply types/index.d.ts]: - `files` are the files in the set, that are being processed
- `origins` is a list of their file paths, to not add duplicates
- `expected` is `origins.length - 1`
- `actual` is the number of done files, which if it matches `expected` signals the file set is done and the engine can continue with writing files and reporting and such
- `pipeline` is a trough pipeline called with completers, which are a way to “plug into” the file set when all files are done running
- `plugins` are a list of those used plugins

---

Fileset could be much better (uniflow?). I’d say `origins`, `expected`, `actual`, and `pipeline` are “hidden internals”. The only thing that uses the fileset [remark-validate-links](https://github.com/remarkjs/remark-validate-links/blob/master/lib/index.js).

File: package.json [hunk 8]
```diff
@@ -38,22 +45,29 @@
...
     "tape": "^4.0.0",
     "unified": "^8.0.0",
+    "vfile": "^4.0.0",
     "vfile-reporter-json": "^2.0.0",
     "vfile-reporter-pretty": "^4.0.0",
     "xo": "^0.28.0"
   },
   "scripts": {
-    "format": "remark . -qfo && prettier --write \"**/*.js\" && xo --fix",
+    "format": "remark . -qfo && prettier --write \"**/*.{js,ts}\" && xo --fix --ignore types",
```
Comment (wooorm) [inline package.json]: Is `--ignore types` the way we should go for other projects with types too?

File: types/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+    /**
+     * Filepath to a configuration file to load
+     */
+    rcPath?: string
+
+    /**
+     * Configuration for the parser and compiler of the processor
+     */
+    settings?: P
```
Comment (wooorm) [inline types/index.d.ts]: I believe this is the same as unified’s API `settings`?

File: types/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+
+    /**
+     * Plugins to use
+     */
+    plugins?: Pluggable<[Settings?], P> | PluggableList<P>
+
+    /**
+     * Optional prefix to use when searching for plugins
+     */
+    pluginsPrefix?: string
```
Comment (wooorm) [inline types/index.d.ts]:
```suggestion
    pluginPrefix?: string
```

File: types/index.d.ts [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+     * Transform config files from a different schema
+     */
+    configTransform?: (config: unknown) => Partial<Options<P, V>>
+
+    /**
+     * Reporter to use
+     *
+     * @defaultValue `require('vfile-reporter')`
+     */
+    reporter?: VFileReporter<V>
```
Comment (wooorm) [inline types/index.d.ts]: A string such as `pretty` or `vfile-reporter-pretty` can also be given

File: types/index.d.ts (hunk 5 above)
Comment (Rokt33r) [reply types/index.d.ts]: `VFile` looks okay but `VFileReporter` should be defined in `vfile-reporter`. Also it is one of deps of this package.

File: package.json (hunk 8 above)
Comment (ChristianMurphy) [reply package.json]: Potentially, XO was getting tripped up by index being a definition file, plus we have dtslint validating definition files.

File: types/index.d.ts (hunk 5 above)
Comment (ChristianMurphy) [reply types/index.d.ts]: I thought about that, what stopped me (maybe I'm over thinking it) is that the VFileReporter type is independent of vfile-reporter.
vfile-reporter happens to be the default here, but it doesn't have to be.
That is why I was maybe thinking of moving it up to vfile, since the concept may be more general.

File: types/index.d.ts (hunk 6 above)
Comment (ChristianMurphy) [reply types/index.d.ts]: Thanks! I'll remove the properties intended to be internal only from the public API.

Comment (wooorm) [reply types/index.d.ts]: Whoops, `plugins` should also be private I think!

Comment (ChristianMurphy) [reply types/index.d.ts]: that leaves, `files`.
Is files here the same as `files` returned directly on the `CallbackContext`?
Should all of `FileSet` be private?

Comment (wooorm) [reply types/index.d.ts]: Hmm. `unified-args` does use `origin`s for its watching capabilities. But the two are heavily coupled so that doesn’t really make it public...

The only really public parts are those documented in [`plugins.md`](https://github.com/unifiedjs/unified-engine/blob/master/doc/plugins.md#contents)

File: types/index.d.ts (hunk 9 above)
Comment (ChristianMurphy) [reply types/index.d.ts]: I think it is?
This is using a generic, by default it is set to `unified.Settings`
https://github.com/unifiedjs/unified-engine/blob/4008651f29abb51154354e1d48c1852261bd60e4/types/index.d.ts#L9-L15
However, because it is generic, if a processor defines more specific settings, those would take precendence.
For example, remark defines `gfm` as a setting, but does not define `notARealSetting`, because of the generic TypeScript warns when `notARealSetting` is used:

https://github.com/unifiedjs/unified-engine/blob/4008651f29abb51154354e1d48c1852261bd60e4/types/unified-engine-tests.ts#L14-L29

File: types/index.d.ts (hunk 10 above)
Comment (ChristianMurphy) [reply types/index.d.ts]: Thanks, updated

File: types/index.d.ts (hunk 11 above)
Comment (ChristianMurphy) [reply types/index.d.ts]: good catch, updated

File: types/index.d.ts (hunk 9 above)
Comment (wooorm) [reply types/index.d.ts]: Perfect!

File: types/index.d.ts (hunk 6 above)
Comment (ChristianMurphy) [reply types/index.d.ts]: Understood, updated FileSet to match the plugins.md description

File: types/index.d.ts [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+  interface Completer {
+    (fileSet: FileSet, next?: CompleterNext): Error | Promise<void>
+
+    /**
+     * Plugins specified through various mechanisms are attached to a new processor for each file.
+     * If a completer is used multiple times, it is invoked multiple times as well.
+     * To prevent completers from attaching multiple times, specify a `pluginId`.
+     * This will ensure only one completer per `pluginId` is added.
+     */
+    pluginId: string
```
Comment (ChristianMurphy) [inline types/index.d.ts]: Is `pluginId` required or optional?

Comment (wooorm) [reply types/index.d.ts]: Optional!

File: types/index.d.ts (hunk 1 above)
Comment (ChristianMurphy) [reply types/index.d.ts]: Thanks updated to be a `Partial<Preset>`

File: types/index.d.ts (hunk 12 above)
Comment (ChristianMurphy) [reply types/index.d.ts]: Thanks, updated

File: types/index.d.ts (hunk 5 above)
Comment (Rokt33r) [reply types/index.d.ts]: Moving `VFileReporter` to `v-file` sounds good to me.
